Don't use getattr in _get_parent_attribute to avoid recursion issues

jimi-c opened the pull request:
Fixes #23609

jimi-c committed:
Don't use getattr in _get_parent_attribute to avoid recursion issues

Fixes #23609

jimi-c commented:
!bot_skip

bcoca commented on lib/ansible/playbook/block.py:
```diff
@@ -295,7 +295,10 @@ def _get_parent_attribute(self, attr, extend=False, prepend=False):
+                        if hasattr(self._parent, '_get_parent_attribute'):
+                            parent_value = self._parent._get_parent_attribute(attr, extend=extend, prepend=prepend)
+                        else:
+                            parent_value = self._parent._attributes.get(attr, None)
```
you need to skip value from parent if include_tasks/include_role, but still inherit

mattclay commented:
CI failure in most integration tests.

jctanner commented:
This terrible patch will get the integration tests to pass ...

```
]$ diff ansible.ansible/lib/ansible/playbook/task.py ansible.patched.pr/lib/ansible/playbook/task.py      
80d79                           
<     _name = FieldAttribute(isa='string', default='')           
427a427,435                     
>                               
>                         if attr == 'environment':              
>                             parent_value = self._parent._get_attr_environment()                                                  
>                         elif attr == 'tags':                   
>                             parent_value = self._parent.tags[:]                                                                  
>                         elif attr == 'no_log':                 
>                             parent_value = self._parent.no_log 
>                         elif attr == 'when':                   
>                             parent_value = getattr(self._parent, attr)
```

Hopefully that helps point out what attributes are missing from the inheritance code.

jimi-c committed:
Move extend/prepend to field attribute

Also removes _get_attr* methods that were basically just calling
_get_parent_attribute because it needed to set those params.

Also modifies _get_parent_attribute() to pull those values from the
FieldAttributes instead of using the ones passed into the function.

jimi-c committed:
Better fixes for _get_parent_attribute
